prov/sockets/src/sock_progress.c

```diff
@@ -2333,29 +2332,27 @@ void sock_pe_remove_rx_ctx(struct sock_rx_ctx *rx_ctx)
 static int sock_pe_progress_rx_ep(struct sock_pe *pe, struct sock_ep_attr *ep_attr,
 					struct sock_rx_ctx *rx_ctx)
 {
-	int ret = 0, i, fd, num_fds;
+	int ret = 0, i, num_fds;
 	struct sock_conn *conn;
 	struct sock_conn_map *map;
+	void *ep_contexts[SOCK_EPOLL_WAIT_EVENTS];
 
 	map = &ep_attr->cmap;
 
-        if (!map->used)
-                return 0;
+	if (!map->used)
+		return 0;
 
-        num_fds = sock_epoll_wait(&map->epoll_set, 0);
-        if (num_fds < 0 || num_fds == 0) {
-                if (num_fds < 0)
-                        SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
-                return num_fds;
-        }
+	num_fds = fi_epoll_wait(map->epoll_set, ep_contexts,
+			SOCK_EPOLL_WAIT_EVENTS, 0);
+	if (num_fds < 0 || num_fds == 0) {
+		if (num_fds < 0)
+			SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
+		return num_fds;
+	}
 
 	fastlock_acquire(&map->lock);
 	for (i = 0; i < num_fds; i++) {
-		fd = sock_epoll_get_fd_at_index(&map->epoll_set, i);
-		if (fd == -1) /* failed to lookup fd due to connection failures */
-			continue;
-
-		conn = ofi_idm_lookup(&ep_attr->conn_idm, fd);
+		conn = ep_contexts[i];
```

Comment:
Ok, let me decompose the problem since fi_epoll_wait() has different implementations depending on if the system supports epoll.
- If epoll_wait() is used directly, I don't think the number of maximum events matters: if an event cannot be retrieve during the first call to epoll_wait(), the next calls to epoll_wait() will return the remaining events. I couldn't find anything to justify my statement in the man for epoll(7) or epoll_wait(2), but that's how I always used epoll_wait().
I think that one of the reason to set the maximum events to more than 1 is for efficiently: you can return multiple events with one single syscall and amortize the overhead induced by repetitive calls to epoll_wait().
- For the implementation of fi_epoll_wait() from src/common.c (based on `poll`), I think that indeed, it's bogus in case `max_contexts` is lower than the number of file descriptors reported by `ep->nfds`: in this case, there is a possibility of losing events if a call `poll` does not consume all the events and a next call to `poll` follows.

I think that the current code...
```
ret = poll(ep->fds, ep->nfds, timeout);  
for (i = ep->index; i < ep->nfds && found < max_contexts; i++) { 
    if (ep->fds[i].revents) {                                    
        contexts[found++] = ep->context[i];                      
        ep->index = i;                                           
    }                                                            
}                                                                
for (i = 0; i < ep->index && found < max_contexts; i++) {        
    if (ep->fds[i].revents) {                                    
        contexts[found++] = ep->context[i];                      
        ep->index = i;                                           
    }                                                            
}
return found;
```

... should be rewritten to something like this (`poll` is called only when all the events from the previous call to `poll` are consumed) :

```
for (i = ep->index; i < ep->nfds && found < max_contexts; i++) { 
    if (ep->fds[i].revents) {                                    
        contexts[found++] = ep->context[i];                      
        ep->index = i;                                           
    }                                                            
}     

if (found >= max_contexts)
    return found;

ret = poll(ep->fds, ep->nfds, timeout);                                                             
for (i = 0; i < ep->index && found < max_contexts; i++) {        
    if (ep->fds[i].revents) {                                    
        contexts[found++] = ep->context[i];                      
        ep->index = i;                                           
    }                                                            
}
return found;